AI/src/search/mcts/nodes/PNMCTSNode.java

```diff
@@ -0,0 +1,339 @@
+package search.mcts.nodes;
+
+import other.RankUtils;
+import other.context.Context;
+import other.move.Move;
+import search.mcts.MCTS;
+
+/**
+ * Node for PN-MCTS tree.
+ * 
+ * @author Dennis Soemers
+ */
+public final class PNMCTSNode extends DeterministicNode
+{
+	
+	//-------------------------------------------------------------------------
+	
+	/**
+	 * Nodes types in search trees in PN-MCTS
+	 * 
+	 * @author Dennis Soemers
+	 */
+	public enum PNMCTSNodeTypes 
+	{
+        /** An OR node */
+        OR_NODE,
+
+        /** An AND node */
+        AND_NODE
+    }
+	
+	/**
+	 * Values of nodes in search trees in PN-MCTS
+	 * 
+	 * @author Dennis Soemers
+	 */
+	public enum PNMCTSNodeValues
+	{
+		/** A proven node */
+		TRUE,
+		
+		/** A disproven node */
+		FALSE,
+		
+		/** Unknown node (yet to prove or disprove) */
+		UNKNOWN
+	}
+	
+	//-------------------------------------------------------------------------
+	
+	/** Proof number for this node */
+	protected double proofNumber;
+	/** Disproof number for this node */
+	protected double disproofNumber;
+	
+	/** The player to move in the root node of the tree this node is in. */
+	protected int rootPlayer;
+	
+	/** what type of node are we? */
+	protected PNMCTSNodeTypes type;
+	
+	/** The value (in terms of proven/disproven/dont know) for this node */
+	protected PNMCTSNodeValues proofValue;
+	
+	/** Are the cached PNS-based terms of childrens' selection scores outdated? */
+	protected boolean childSelectionScoresDirty = false;
+	
+	/** Cached PNS-based terms for selection scores for all our children (including unexpanded ones) */
+	protected final double[] childrenPNSSelectionTerms;
```

Comment:
I stored all these terms for all children together in the parent node.  This lets me also store PNS-terms for children that do not exist yet.